tests/unittests/core/fileio/fileutilstest.cpp

```diff
@@ -0,0 +1,313 @@
+/*
+ * LibrePCB - Professional EDA for everyone!
+ * Copyright (C) 2013 LibrePCB Developers, see AUTHORS.md for contributors.
+ * https://librepcb.org/
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*******************************************************************************
+ *  Includes
+ ******************************************************************************/
+
+#include <gtest/gtest.h>
+#include <librepcb/core/exceptions.h>
+#include <librepcb/core/fileio/filepath.h>
+#include <librepcb/core/fileio/fileutils.h>
+#if defined(Q_OS_WIN32) || defined(Q_OS_WIN64)  // Windows
+#include <fileapi.h>
+#endif
+
+#include <QtCore>
+
+/*******************************************************************************
+ *  Namespace
+ ******************************************************************************/
+namespace librepcb {
+namespace tests {
+
+/*******************************************************************************
+ *  Test Class
+ ******************************************************************************/
+
+class FileUtilsTest : public ::testing::Test {
+protected:
+  FilePath root;
+  FilePath rootFile;  // source for all operations
+  FilePath rootFileHidden;
+
+  FilePath subdir;
+  FilePath subdirFile;
+  FilePath subdirSubdir;
+  FilePath subdirSubdirFile;
+  FilePath subdirSubdirFileHidden;
+
+  FilePath rootFileMissing;
+  FilePath rootFileCopy;
+  FilePath subdirCopy;
+  FilePath subdirCopyFile;
+  FilePath subdirCopySubdir;
+  FilePath subdirCopySubdirFile;
+  FilePath subdirCopySubdirFileHidden;
+
+  QStringList filter;
+
+  static void SetupFile(const FilePath& pth, const QByteArray& content) {
+    QFile f(pth.toNative());
+    if (f.open(QIODevice::ReadWrite | QIODevice::NewOnly)) f.write(content);
```

Comment:
Good point is not managing fail inside helper function. It is nice facepalm i got and accepting it :-)

The second is nice point from the test perspective. Test on another technology than it runs can reveal more bugs. My previous assumption is "use app codebase that is already used" so when You use Qt everywhere, I should use Qt here too, to be consistent. This assumption seems to be wrong.

In `c++17` there is `std::filesystem` as platform independent native and with that lot of filesystem code in librepcb can be outdated but yeah, i know you don't like it, there *is* an reason by now so I should wait ... :-) 

So the question is I should rewrite that by using:

* plain C `fread`/`fwrite`
* C++ `std::fstream`
* Qt `QFile`

Reply:
> My previous assumption is "use app codebase that is already used" so when You use Qt everywhere, I should use Qt here too, to be consistent. This assumption seems to be wrong.

It's not wrong, generally of course we use Qt everywhere to keep the code as portable as possible :slightly_frowning_face: The `FileUtils` test is a bit a special case since we're not just testing logic inside our application - in fact, the function under test forwards data to the operating system (leaving our application) and the test reads back the data from the operating system. So the behavior of the operating system (with its filesystem) is actually also verified by the test, which is very unique for the `FileUtils` test (only very few other test are involving the operating system). Just because of this, it came in my mind that using a different API for the readback might make sense, to hopefully cover issues related to Qt's file functions. For (theoretical) example, if Qt cached read/write operation, a readback might be successful even though the data has not been written to the file system yet.

> So the question is I should rewrite that by using:

I'd give `std::fstream` a try. I don't like using C functions if not absolutely necessary, as they are much more cumbersome and errorprone than the C++ API (e.g. error handling is usually simpler with C++).

However, it is just an idea an maybe it's not as simple as it sounds. So generally I'd also be fine using `QFile` for verification. For example if `std::fstream` turns out to be difficult, don't waste time with it and just stick with `QFile`. In the end, Qt developers should have tested `QFile` already :see_no_evil: